tests/store/artifact/test_ftp_artifact_repo.py

```diff
@@ -211,32 +211,99 @@ def test_log_artifact_multiple_calls(ftp_mock, tmpdir):
     assert ftp_mock.storbinary.call_args_list[0][0][0] == "STOR test2.txt"
 
 
+def __posixpath_parents(pathname, root):
+    parents = []
+    parent, _ = posixpath.split(pathname)
+    root = posixpath.normpath(root)
+    parents.append(parent)
+    while parent != "/" and parent != root:
+        parent, _ = posixpath.split(parent)
+        parents.append(parent)
+    return parents
+
+
 @pytest.mark.parametrize("artifact_path", [None, "dir", "dir1/dir2"])
 def test_log_artifacts(artifact_path, ftp_mock, tmpdir):
-    repo = FTPArtifactRepository("ftp://test_ftp/some/path")
+    # Setup FTP mock.
+    dest_path_root = "/some/path"
+    repo = FTPArtifactRepository("ftp://test_ftp" + dest_path_root)
 
     repo.get_ftp_client = MagicMock()
     call_mock = MagicMock(return_value=ftp_mock)
     repo.get_ftp_client.return_value = MagicMock(__enter__=call_mock)
 
+    dir_created = set([dest_path_root])
+    file_created = set()
+    cwd_history = []
+
+    def mkd_mock(pathname):
+        assert pathname == posixpath.abspath(pathname), "mock not implemented"
+        parent, _ = posixpath.split(pathname)
+        if parent not in dir_created:
+            raise ftplib.error_perm
+        assert pathname not in dir_created, "mock not implemented"
+        dir_created.add(pathname)
+
+    ftp_mock.mkd = MagicMock(side_effect=mkd_mock)
+
+    def cwd_mock(pathname):
+        assert pathname == posixpath.abspath(pathname), "mock not implemented"
+        if pathname not in dir_created:
+            raise ftplib.error_perm
+        cwd_history.append(pathname)
+
+    ftp_mock.cwd = MagicMock(side_effect=cwd_mock)
+
+    def storbinary_mock(cmd, _):
+        head, basename = cmd.split(" ", 1)
+        assert head == "STOR"
+        assert "/" not in basename
+        assert len(cwd_history) > 0, "mock not implemented"
+        file_created.add(posixpath.join(cwd_history[-1], basename))
+
+    ftp_mock.storbinary = MagicMock(side_effect=storbinary_mock)
+
+    # Test
     subd = tmpdir.mkdir("data").mkdir("subdir")
     subd.join("a.txt").write("A")
     subd.join("b.txt").write("B")
     subd.join("c.txt").write("C")
+    subd.mkdir("empty1")
+    subsubd = subd.mkdir("subsubdir")
+    subsubd.join("aa.txt").write("AA")
+    subsubd.join("bb.txt").write("BB")
+    subsubd.join("cc.txt").write("CC")
+    subsubd.mkdir("empty2")
+
+    dest_path = (
+        dest_path_root if artifact_path is None else posixpath.join(dest_path_root, artifact_path)
+    )
+    dir_expected = set(
```

Comment:
```suggestion
    dirs_expected = set(
```

to indicate this variable represents a collection of directories?